include/openssl/hpke.h

```diff
@@ -0,0 +1,466 @@
+ *
+ * In all these APIs public keys are passed as buffers whereas
+ * private keys as passed as EVP_PKEY pointers.
+ */
+int OSSL_HPKE_CTX_set1_authpub(OSSL_HPKE_CTX *ctx,
+                               unsigned char *pub,
+                               size_t publen);
+
+/**
+ * @brief ask for the state of the sequence of seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq returns the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The value returned is the next one to be used when sealing
+ * or opening (so as we start at zero this will be 1 after the
+ * first successful call to seal or open)
+ *
+ * seq is a uint64_t as that's what two other implementations
+ * chose
+ */
+int OSSL_HPKE_CTX_get0_seq(OSSL_HPKE_CTX *ctx, uint64_t *seq);
+
+/**
+ * @brief set the sequence value for seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq set the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The next seal or open operation will use this value.
+ */
+int OSSL_HPKE_CTX_set1_seq(OSSL_HPKE_CTX *ctx, uint64_t seq);
+
+/**
+ * @brief sender seal function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param ct is the ciphertext output
+ * @param ctlen is the size the above
+ * @param pub is the recipient public key octets
+ * @param publen is the size the above
+ * @param recip is the EVP_PKEY form of recipient public value
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @param aad is the aad parameter
+ * @param aadlen is the size the above
+ * @param pt is the plaintext
+ * @param ptlen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * This can be called once, or multiple, times.
+ *
+ * If no KEM private key has been set in the context an ephemeral
+ * key will be generated and used for the duration of the context.
+ *
+ * The ciphertext buffer (ct) should be big enough to include
+ * the AEAD tag generated from encryptions and the ``enc`` buffer
+ * (the ephemeral public key) needs to be big enough for the
+ * relevant KEM. ``OSSL_HPKE_expansion`` can be used to determine
+ * the sizes needed.
+ */
+int OSSL_HPKE_sender_seal(OSSL_HPKE_CTX *ctx,
+                          unsigned char *enc, size_t *enclen,
+                          unsigned char *ct, size_t *ctlen,
+                          unsigned char *pub, size_t publen,
+                          const unsigned char *info, size_t infolen,
+                          const unsigned char *aad, size_t aadlen,
+                          const unsigned char *pt, size_t ptlen);
+
+/**
+ * @brief recipient open function
+ * @param ctx is the pointer for the HPKE context
+ * @param pt is the plaintext
+ * @param ptlen is the size the above
+ * @param enc is the sender's ephemeral public value
```

Comment:
It should say the encoded public key..

Reply:
for the curves we use now, yes. apparently in general, no - I'm told the ephemeral value for some potential future schemes could be something other than an encoded public key (but I don't care much:-)